### ISSUE TITLE
`PostUpdateRepository::updateOrder()` direct database operation doesn't trigger WP hooks

### ISSUE THREAD

**Issue by dhuyck** [kylephillips/wp-nested-pages#273]
When sorting pages, and especially when changing a page's parent, the `post_updated` action is not triggered. The `PostUpdateRepository::updateOrder()` method does direct database manipulation, so `wp_insert_post()` is never called, and its hooks are never fired.

This becomes a problem when using redirection plugins that create redirects to a page's new URL. Because the `post_updated` hook is never triggered, the change goes undetected, resulting in 404s.

**Comment by dhuyck** [comment]
I tried to use the hooks fired in the `PostUpdateRepository::updateOrder()` method (`nestedpages_post_order_updated` and `nestedpages_posts_order_updated`), but there is not enough data there to create the before/after post objects, since this is a straight `UPDATE` SQL query. Without being able to calculate the page's URL before and after the change, the redirection plugin we are using cannot figure out a proper redirect to add.

**Comment by timotheemoulin** [comment]
I also refresh my redirections based on the `nestedpages_post_order_updated` post (so I can target a single post).

What kind of data are you missing there?

You've got the `$post_id`, `$parent`, and `$key` (menu name).

**Comment by dhuyck** [comment]
> What kind of data are you missing there?

I think what I ran into was not being able to see what the previous permalink was - after the post is changed directly in the data (and not through a wp_update_post() call) there is no way to reconstruct the URL before the change, so we can't say "previous URL should redirect to current URL".

**Comment by timotheemoulin** [comment]
Indeed,
I think we might need to add all the modified value in the hook action.
Would something like that be sufficient?

```php
do_action('nestedpages_post_order_updated', $post_id, $parent, $key, $post);
```
As we already retrieve the original post during the `for` loop, there is no new DB query needed.

**Comment by timotheemoulin** [comment]
Would something like that be sufficient to your needs? As the post object is not loaded, I don't think that would be a good idea to do some more DB queries (the loop already save every post one at the time).

```php

foreach ($posts as $key => $post)
{
...

do_action('nestedpages_post_order_before_update', $post_id, $parent, $key);

$query = ...
do_action('nestedpages_post_order_updated', $post_id, $parent, $key);
...
```

**Comment by dhuyck** [comment]
[I wrote a PR](https://github.com/kylephillips/wp-nested-pages/pull/277) that fixes this for (at least) our context - the redirect plugin we use hooks into the 'post_updated' action, so updating the post "properly" made that work for us.

Here's the relevant bit, lines 55-62:
```
if ( wp_get_post_parent_id( $post_id ) !== $parent ) {
	wp_update_post( [
		'ID'          => $post_id,
		'post_parent' => $parent,
		'menu_order'  => $key,
		'post_type'   => get_post_type( $post_id ),
	] );
} else {
	// ...do the existing Nested Pages way ...
}
```

Performance is already not-great on larger sites, as it loops through each page, but I'm not sure there is a better way (other than throwing more CPU at it). If it could be run asynchronously, that would be cool, but I'm not aware of a simple way to do so.

**Comment by kylephillips** [comment]
Fixed in 0db6bfa99ee9c68f2305b5e01a7feca228faf048
